# Results

## Alignment

In order to use these data to build our phylogenies, we need to align each gene.
This allows our tree-building software to compare species using nucleotides that we believe share ancestry. 
We will use the software MAFFT.
Due to the nature of this server different programs are in different "environments" so you'll need to deactivate the hybpiper environment and activate one for mafft.

```
conda deactivate
conda activate /mnt/homes4celsrs/shared/envs/mafft
```

Now MAFFT is available to run.
However, before we align our data let's take a look at the output.
Do you remember how to print out particular lines of a file?
Try printing all of the lines that start with `'>'` in a particular FNA file. **Make sure to use `'>'` including the single quotes when searching.**
You should notice that HybPiper has added some information to these lines so they include more than just the name of the sample. Because these additions differ across genes subsequent analyses may treat them as separate samples.

Before we do the alignment we need to remove this extra information.
Again, we can think about the pattern we are looking for: we want to go through each FNA file (this should suggest using a loop) and keep the first "word" in each line (think of a word as a set of characters not separated by spaces).
We use the `cut` command to divide each line of our file into columns.
We use the `-f` flag to indicate the field (column) we want to keep, and the `-d` flag to indicate how to separate columns (i.e. with a space: `' '`).

An example cut command for one file could look like the following:

```
cut -f 1 -d ' ' 7577.FNA > 7577.FNA.fa
```

Now use a loop as before to repeat this command for all FNA files.

```{bash, echo=FALSE, eval=FALSE}
for f in *FNA; do
cut -f 1 -d ' ' $f > ${f}.fa
done
```

Now we can loop through each of these files and output an alignment.
The basic mafft command looks like the following, assuming you replace FILE with a particular file.

```
mafft --auto --thread 1 7577.FNA.fa > 7577.FNA.fa.fasta
```

Remember you have 350 genes (i.e. 350 fasta files) so you want to run these alignments in a loop.
As before, write a loop to go through each FNA.fa file and output an alignment.


```{bash, echo=FALSE, eval=FALSE}
for F in *FNA.fa; do
mafft --auto --thread 1 $F > ${F}.fasta
done
```

## View in R

The best way to view an alignment is in a specialized program, but because we have R
easily available we'll view an example here.

Make a new script to run this alignment view.
You should first load the `ape` and `ggmsa` library. (This uses the gg as in ggplot and msa, which stands for multiple sequence alignment.)

* Read in your fasta file with ape's `read.FASTA`.
* Use the ggmsa command as follows, substituting the particular variable that contains your fasta file. Allow a little while for this to run. You may need to click the Zoom button above the plot to see it more clearly.

```
ggmsa(seqs_7577, start = 20, end = 120, char_width = 0.5, seq_name = T, color = "Chemistry_NT") +
  geom_msaBar()
```

Now repeat this process for the same gene with the unaligned data. Can you see the difference?

```{bash, echo=FALSE, eval=FALSE}
library(ggmsa)  #use R 4.1.2
library(ape)

#view aligned
tax1_seqs_4471 <- ("AndesWorkshop2023/TaxonSet1/353reads/4471.FNA.fasta.fa")
ggmsa(tax1_seqs_4471, start = 20, end = 120, char_width = 0.5, seq_name = T, color = "Chemistry_NT") + 
  geom_msaBar() #this is slow - wait!

#view unaligned
tax1_seqs_4471u <- ("AndesWorkshop2023/TaxonSet1/353reads/4471.FNA")
ggmsa(tax1_seqs_4471u, start = 20, end = 120, char_width = 0.5, seq_name = T, color = "Chemistry_NT") + 
  geom_msaBar() #this is slow - wait!
```

## Trimal

TBD
```{bash, echo=FALSE, eval=FALSE}
https://hackmd.io/@mossmatters/Sy6md0prY
```

## Concatenated data analysis

There are multiple ways to analyze data.
The first is to concatenate everything.
Make sure you are in your Terminal.

```
conda deactivate
conda activate /mnt/homes4celsrs/shared/envs/amas

python3 /mnt/homes4celsrs/shared/envs/amas/bin/AMAS.py concat -f fasta \
-d dna --out-format fasta --part-format raxml -i *FNA.fa.fasta \
-t concatenated.fasta -p partitions.txt

```
Note that if some data was not found for some genes you may need to delete files (use the `rm` command) to get AMAS to concatenate your data.


### Building trees with IQTree

We will build our first tree using our complete concatenated dataset.
The program IQtree infers phylogenetic trees by maximum likelihood.
This approach starts with a tree and calculates the likelihood of the data on the tree (i.e. calculating the probability of each site fitting the tree given a model of substitution and multiplying them together).
We use the General Time Reversible (GTR) model to allow sites to change back and forth among different bases with particular probabilities.
We also allow a Gamma (G) distribution of rates of substitution across sites.
We allow partitioning of the data by gene so that different genes can evolve according to different models.
Additionally we have included bootstrapping in our analysis to get a measure of support for relationships.

```
conda deactivate
conda activate /mnt/homes4celsrs/shared/envs/iqtree

iqtree2 -nt 2 -s concatenated.fasta -spp partitions.txt -pre iqtree_tree -B 1000 -m GTR+G
```

### Viewing trees

* Make a new RScript to view your tree in R.
* Load the libraries tidyverse and ape.
* Use the `read.tree` command and provide the path to the tree as the argument
* Use the plot command providing the tree variable as the argument, then add `, show.node.label = TRUE`

```{r, eval=FALSE, echo=FALSE}

tax1 <- read.tree("AndesWorkshop2023/TaxonSet1/353reads/iqtree_tree.treefile")
plot(tax1, show.node.label = TRUE)
```

#### Rooting your tree

Our tree-building programs create trees that are unrooted because we do not know the direction of changes among species (e.g. for a single difference we are unable to say if an A mutated to a T or vice versa).
Thus, we should properly view out trees as unrooted.

```
plot(tax1, "u", show.node.label = TRUE, cex = .5)
```

Note that I have added an argument to make the font size  bit smaller so we can read all of the labels.
You should adjust this value as needed and use the Zoom feature to better view your tree.

However, we have created taxon sets where one of the included species is known to be more distantly related.
We can use this to set the root for the tree.
You should view the list of taxa for your group in the file in the shared folder for this workshop.
The last taxon in your list is the outgroup.
You can root your tree as in the following example using the `root` command and the outgroup argument.
Note that your outgroup will be different if you are using a different taxon set.
Additionally you must use the tip label that you can see on your tree not the labels in this spreadsheet.
If you want to view the tip labels as a list use

```
tax1$tip.label
tax1_root <- root(tax1, outgroup = "A217_CKDN220062756-1A")
```

Now that you have rooted your tree you can plot this new tree as before.

#### Relabeling your tips with species information

There are a couple of ways to relabel the tips of your trees with correct species names.
We will use a straightforward, manual approach.
We will look at the list of tips and then manually replace them with a list of species names.
You should keep in mind that this approach is prone to error and challenging for many taxa; however an automated approach is a bit more challenging to set up in this course.

* List the tip labels as you did in the previous step
* Make a list of new tip labels - it should look something like the following but with more and different species

```
sp_names <- c("Centropogon mandonis", "Siphocampylus andinus", "Centropogon mandonis")
```

* Assign these species names to the tip labels

```
tax1$tip.label <- sp_names
```

If you relabeled the tips of your unrooted tree you need to root the tree with the appropriate outgroup name. You can then plot this rooted and corrected tree.

### Tree for all data using output of individual hybpiper runs

In the prior analysis we examined all of the data combined.
For the next analysis we will estimate individual gene trees and then combine these into a species tree.

```
iqtree -s concatenated.fasta -S partitions.txt -pre iqtree.loci -nt 2
```

-S tells IQ-TREE to infer separate trees for each partition.
The output files are the same, except that now your treefile will contain a set of gene trees.

For a more extensive tutorial on IQTree see
http://www.iqtree.org/workshop/molevol2019

## Species tree analyses

An alternative is to use a species tree approach.
We will use the software ASTRAL.

```
java -jar /mnt/homes4celsrs/shared/ASTRAL/astral.5.7.8.jar -i iqtree.loci.treefile -o astral_output.tre 2>astral.log
```

* -i is the flag for the input file
* -o is the flag for the output file
* 2> saves the output log information

The output in is Newick format and gives:

* the species tree topology
* branch lengths in coalescent units (only for internal branches or for terminal branches if that species has multiple individuals)
* branch supports measured as local posterior probabilities

* View this tree as before and compare it to the other tree you estimated using IQTree for partitioned concatenated data.

More information on astral can be found at
https://github.com/smirarab/ASTRAL

## Support for relationships

In our initial IQTree analysis we obtained bootstrap support for each node.
An alternative approach are gene and site concordance factors.
For more information see http://www.robertlanfear.com/blog/files/concordance_factors.html .
You can compute gCF and sCF for the tree inferred under the partition model:

```
iqtree -t iqtree_tree.treefile --gcf iqtree.loci.treefile -s concatenated.fasta --scf 100 -nt 2
```

* -t specifies a tree
* --gcf specifies the gene-trees file
* --scf 100 to draw 100 random quartets when computing sCF.

* Repeat for the astral tree if it differs.
* View the tree (iqtree_tree.treefile.cf.tree) in R with these support values. The tree will show boostrap/gcf/scf on the nodes.

## MrBayes

The next approach we will use for tree building is implemented in MrBayes.
We need to do some setup before we can run this software.
`R` has some tools we can use to convert out data to the write format and add some instructions.
Note that some of the following was adapted from https://gtpb.github.io/MEVR16/bayes/mb_example.html

* Load your data into R (use a new script for organizational purposes)

```
library(ape)
library(tidyverse)

myseqs <- read.dna("results/concatenated.fasta",format="fasta",as.matrix=FALSE)
```

MrBayes requires Nexus format, with an added block giving instructions to MrBayes. We first save the data as Nexus format, and read back in to manipulate further.

```
write.nexus.data(as.character(myseqs),"results/concatenated.nex",interleaved=TRUE,gap="-",missing="N")
myseqs_nex <- readLines("results/concatenated.nex")
```

Fix missing because AMAS using ? for missing

```
myseqs_nex <- gsub('\\?','n',myseqs_nex)
```

First execution block

```
mbblock1 <- "
begin mrbayes;
  set autoclose=yes;
"
```

Second execution block has information about partitions.
We get this from the partitions file generated by AMAS but we have to do some reformatting.

* Read the file with `read.table`
* Get just the last column using the `pull` function
* Now you can convert this list to a string to be inserted into the output file

```
partition_string <- toString(partitions,sep = ", ")
```

Finally, put all the information necessary for MrBayes into a single string

```
mbblock2 <- paste0("  partition favored = ",nrow(partition_file),":",partition_list,";")
```

Add a block for the MCMC parameters.

```
mbblock2 <- "
  mcmc ngen=10000000 nruns=2 nchains=2 samplefreq=1000;
  sump;
  sumt;
end;
"
```

We then paste the blocks together and write to a file.

```
myseqs_nexus_withmb <- paste(paste(myseqs_nex,collapse="\n"),mbblock1,mbblock2,sep="")
write(myseqs_nexus_withmb,file="concatenated.nex.mb")
```

Now run (on the command line) (probably in background as you did previously).
```
../../shared/bin/mb "results/concatenated.nex.mb"
```

Before looking at your tree check the output. In some cases you may need to run your analysis for longer and a note to this effect will be toward the end of the output. Make sure to give the log file for this run a relevant name. You don't want to overwrite the output from HybPiper and you might want to keep this output if you do a longer run. Additionally, if you rerun the analysis with additional time make sure you don't overwrite the current tree output.

You can find your tree in your results folder with the extension `.con.tre`.
Because Mr. Bayes generates output in a particular format you will need to read it in slightly differently and convert it to the standard `phylo` format.

```
mrbayes_tree <- treeio::read.mrbayes("results/concatenated.nex.mb.con.tre")
mrbayes_phylo <- treeio::as.phylo(mrbayes_tree)
probs = as_tibble(mrbayes_tree) %>% pull(prob_percent)
mrbayes_phylo$node.label = probs[(length(mrbayes_phylo$tip.label)+1):length(probs)]
```

* Reroot your tree
* Fix the names to be species
* Plot your tree

## What do your trees tell you?



