<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 15 Discrete variable modeling | Andes Phylogenomics Workshop</title>
  <meta name="description" content="Step-by-step guide to a project in phylogenomics." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 15 Discrete variable modeling | Andes Phylogenomics Workshop" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Step-by-step guide to a project in phylogenomics." />
  <meta name="github-repo" content="rachelss/Andes_workshop_codebook" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 15 Discrete variable modeling | Andes Phylogenomics Workshop" />
  
  <meta name="twitter:description" content="Step-by-step guide to a project in phylogenomics." />
  

<meta name="author" content="Rachel Schwartz and Rosana Zenil-Ferguson" />


<meta name="date" content="2023-06-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="phylogenetic-generalized-linear-models.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Andes Phylogenomics Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this workshop</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#case-studies"><i class="fa fa-check"></i><b>2.2</b> Case Studies</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#course-objectives"><i class="fa fa-check"></i><b>2.3</b> Course Objectives</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-and-analysis.html"><a href="data-and-analysis.html"><i class="fa fa-check"></i><b>3</b> Data and analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="data-and-analysis.html"><a href="data-and-analysis.html#genomic-data"><i class="fa fa-check"></i><b>3.1</b> Genomic data</a></li>
<li class="chapter" data-level="3.2" data-path="data-and-analysis.html"><a href="data-and-analysis.html#phylogenies"><i class="fa fa-check"></i><b>3.2</b> Phylogenies</a></li>
<li class="chapter" data-level="3.3" data-path="data-and-analysis.html"><a href="data-and-analysis.html#comparative-analyses"><i class="fa fa-check"></i><b>3.3</b> Comparative analyses</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="trait-data.html"><a href="trait-data.html"><i class="fa fa-check"></i><b>4</b> Trait Data</a>
<ul>
<li class="chapter" data-level="4.1" data-path="trait-data.html"><a href="trait-data.html#getting-started-with-r"><i class="fa fa-check"></i><b>4.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="trait-data.html"><a href="trait-data.html#option-1-log-in-to-the-server"><i class="fa fa-check"></i><b>4.1.1</b> Option 1: Log in to the server</a></li>
<li class="chapter" data-level="4.1.2" data-path="trait-data.html"><a href="trait-data.html#option-2-work-on-your-personal-computer"><i class="fa fa-check"></i><b>4.1.2</b> Option 2: Work on your personal computer</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="trait-data.html"><a href="trait-data.html#load-and-set-up-data"><i class="fa fa-check"></i><b>4.2</b> Load and set up data</a></li>
<li class="chapter" data-level="4.3" data-path="trait-data.html"><a href="trait-data.html#beginning-an-analysis"><i class="fa fa-check"></i><b>4.3</b> Beginning an analysis</a></li>
<li class="chapter" data-level="4.4" data-path="trait-data.html"><a href="trait-data.html#data-visualization"><i class="fa fa-check"></i><b>4.4</b> Data Visualization</a></li>
<li class="chapter" data-level="4.5" data-path="trait-data.html"><a href="trait-data.html#statistical-tests-of-relationships"><i class="fa fa-check"></i><b>4.5</b> Statistical tests of relationships</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="computational-skills.html"><a href="computational-skills.html"><i class="fa fa-check"></i><b>5</b> Computational skills</a>
<ul>
<li class="chapter" data-level="5.1" data-path="computational-skills.html"><a href="computational-skills.html#basic-navigation"><i class="fa fa-check"></i><b>5.1</b> Basic navigation</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="getting-your-data.html"><a href="getting-your-data.html"><i class="fa fa-check"></i><b>6</b> Getting your data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="getting-your-data.html"><a href="getting-your-data.html#repeating-analyses"><i class="fa fa-check"></i><b>6.1</b> Repeating analyses</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="tree-thinking-quiz.html"><a href="tree-thinking-quiz.html"><i class="fa fa-check"></i><b>7</b> Tree thinking quiz</a></li>
<li class="chapter" data-level="8" data-path="target-capture-data.html"><a href="target-capture-data.html"><i class="fa fa-check"></i><b>8</b> Target capture data</a>
<ul>
<li class="chapter" data-level="8.1" data-path="target-capture-data.html"><a href="target-capture-data.html#background"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="target-capture-data.html"><a href="target-capture-data.html#running-hybpiper-to-assemble-genes"><i class="fa fa-check"></i><b>8.2</b> Running HybPiper to assemble genes</a></li>
<li class="chapter" data-level="8.3" data-path="target-capture-data.html"><a href="target-capture-data.html#process-outputs"><i class="fa fa-check"></i><b>8.3</b> Process outputs</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="target-capture-data.html"><a href="target-capture-data.html#viewing-information"><i class="fa fa-check"></i><b>8.3.1</b> Viewing information</a></li>
<li class="chapter" data-level="8.3.2" data-path="target-capture-data.html"><a href="target-capture-data.html#obtain-gene-data"><i class="fa fa-check"></i><b>8.3.2</b> Obtain gene data</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="target-capture-data.html"><a href="target-capture-data.html#visual-inspection-of-data"><i class="fa fa-check"></i><b>8.4</b> Visual inspection of data</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="results.html"><a href="results.html"><i class="fa fa-check"></i><b>9</b> Results</a>
<ul>
<li class="chapter" data-level="9.1" data-path="results.html"><a href="results.html#alignment"><i class="fa fa-check"></i><b>9.1</b> Alignment</a></li>
<li class="chapter" data-level="9.2" data-path="results.html"><a href="results.html#view-in-r"><i class="fa fa-check"></i><b>9.2</b> View in R</a></li>
<li class="chapter" data-level="9.3" data-path="results.html"><a href="results.html#concatenated-data-analysis"><i class="fa fa-check"></i><b>9.3</b> Concatenated data analysis</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="results.html"><a href="results.html#building-trees-with-iqtree"><i class="fa fa-check"></i><b>9.3.1</b> Building trees with IQTree</a></li>
<li class="chapter" data-level="9.3.2" data-path="results.html"><a href="results.html#viewing-trees"><i class="fa fa-check"></i><b>9.3.2</b> Viewing trees</a></li>
<li class="chapter" data-level="9.3.3" data-path="results.html"><a href="results.html#tree-for-all-data-using-output-of-individual-hybpiper-runs"><i class="fa fa-check"></i><b>9.3.3</b> Tree for all data using output of individual hybpiper runs</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="results.html"><a href="results.html#species-tree-analyses"><i class="fa fa-check"></i><b>9.4</b> Species tree analyses</a></li>
<li class="chapter" data-level="9.5" data-path="results.html"><a href="results.html#support-for-relationships"><i class="fa fa-check"></i><b>9.5</b> Support for relationships</a></li>
<li class="chapter" data-level="9.6" data-path="results.html"><a href="results.html#mrbayes"><i class="fa fa-check"></i><b>9.6</b> MrBayes</a></li>
<li class="chapter" data-level="9.7" data-path="results.html"><a href="results.html#what-do-your-trees-tell-you"><i class="fa fa-check"></i><b>9.7</b> What do your trees tell you?</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dates-and-evolution.html"><a href="dates-and-evolution.html"><i class="fa fa-check"></i><b>10</b> Dates and Evolution</a>
<ul>
<li class="chapter" data-level="10.1" data-path="dates-and-evolution.html"><a href="dates-and-evolution.html#dates-times-on-trees"><i class="fa fa-check"></i><b>10.1</b> Dates / Times on Trees</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html"><i class="fa fa-check"></i><b>11</b> Understanding the likelihood function</a>
<ul>
<li class="chapter" data-level="11.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#background-1"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-model"><i class="fa fa-check"></i><b>11.2</b> The model</a></li>
<li class="chapter" data-level="11.3" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-likelihood-function"><i class="fa fa-check"></i><b>11.3</b> The likelihood function</a></li>
<li class="chapter" data-level="11.4" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#a-mini-example-with-a-sample-size-of-two"><i class="fa fa-check"></i><b>11.4</b> A mini example with a sample size of two</a></li>
<li class="chapter" data-level="11.5" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-bird-alarm-example"><i class="fa fa-check"></i><b>11.5</b> The bird alarm example</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#calculate-the-mle-for-dl-and-f-and-the-likelihood-value-evaluated-at-the-mle"><i class="fa fa-check"></i><b>11.5.1</b> Calculate the MLE for DL and F and the likelihood value evaluated at the MLE</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#where-is-the-evidence-for-different-behavior"><i class="fa fa-check"></i><b>11.6</b> Where is the evidence for different behavior?</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#plot-the-likelihood-for-dead-leaf-gleaners"><i class="fa fa-check"></i><b>11.6.1</b> Plot the likelihood for dead-leaf gleaners</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#are-the-freezing-times-different-for-the-two-groups"><i class="fa fa-check"></i><b>11.7</b> Are the freezing times different for the two groups?</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html"><i class="fa fa-check"></i><b>12</b> What is Bayesian statistics?</a>
<ul>
<li class="chapter" data-level="12.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#what-about-all-the-fighting-of-frequentists-vs.-bayesians"><i class="fa fa-check"></i><b>12.1</b> What about all the fighting of “frequentists” vs. “Bayesians”</a></li>
<li class="chapter" data-level="12.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#when-do-we-choose-bayesian-statistics"><i class="fa fa-check"></i><b>12.2</b> When do we choose Bayesian statistics?</a></li>
<li class="chapter" data-level="12.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#basic-concepts-of-bayesian-statistics"><i class="fa fa-check"></i><b>12.3</b> Basic concepts of Bayesian Statistics</a></li>
<li class="chapter" data-level="12.4" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#example-online-dating"><i class="fa fa-check"></i><b>12.4</b> Example: Online dating</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#pab-is-difficult-to-estimate"><i class="fa fa-check"></i><b>12.4.1</b> <span class="math inline">\(P(A|B)\)</span> is difficult to estimate</a></li>
<li class="chapter" data-level="12.4.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#coming-back-to-the-initial-idea-of-how-we-define-bayesian-statistics"><i class="fa fa-check"></i><b>12.4.2</b> Coming back to the initial idea of how we define Bayesian statistics</a></li>
<li class="chapter" data-level="12.4.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-random-variable"><i class="fa fa-check"></i><b>12.4.3</b> The random variable</a></li>
<li class="chapter" data-level="12.4.4" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-prior-distribution"><i class="fa fa-check"></i><b>12.4.4</b> The prior distribution</a></li>
<li class="chapter" data-level="12.4.5" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-data-and-their-distribution"><i class="fa fa-check"></i><b>12.4.5</b> The Data and their distribution</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-likelihood-function-1"><i class="fa fa-check"></i><b>12.5</b> The likelihood function</a></li>
<li class="chapter" data-level="12.6" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-update-the-posterior-distribution"><i class="fa fa-check"></i><b>12.6</b> The update! The posterior distribution</a></li>
<li class="chapter" data-level="12.7" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#optional-section-doing-your-first-mcmc"><i class="fa fa-check"></i><b>12.7</b> Optional section: Doing your first MCMC</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-proposal-function"><i class="fa fa-check"></i><b>12.7.1</b> The proposal function</a></li>
<li class="chapter" data-level="12.7.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-metropolis-hastings"><i class="fa fa-check"></i><b>12.7.2</b> The Metropolis-Hastings</a></li>
<li class="chapter" data-level="12.7.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#visualizing-the-mcmc-as-a-way-to-find-out-convergence"><i class="fa fa-check"></i><b>12.7.3</b> Visualizing the MCMC as a way to find out convergence</a></li>
</ul></li>
<li class="chapter" data-level="12.8" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#changing-the-proposal"><i class="fa fa-check"></i><b>12.8</b> Changing the proposal</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html"><i class="fa fa-check"></i><b>13</b> Introduction to macroevolution modeling</a>
<ul>
<li class="chapter" data-level="13.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#reading-dataset-and-phylogenetic-tree"><i class="fa fa-check"></i><b>13.1</b> Reading dataset and phylogenetic tree</a></li>
<li class="chapter" data-level="13.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-a-phylogenetic-tree-with-data"><i class="fa fa-check"></i><b>13.2</b> Plotting a phylogenetic tree with data</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-continuous-data"><i class="fa fa-check"></i><b>13.2.1</b> Plotting continuous data</a></li>
<li class="chapter" data-level="13.2.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-discrete-and-continuous-data"><i class="fa fa-check"></i><b>13.2.2</b> Plotting discrete and continuous data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#introduction-to-brownian-motion"><i class="fa fa-check"></i><b>13.3</b> Introduction to Brownian Motion</a></li>
<li class="chapter" data-level="13.4" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#brownian-motion-in-phylogenetic-comparative-methods"><i class="fa fa-check"></i><b>13.4</b> Brownian motion in phylogenetic comparative methods</a></li>
<li class="chapter" data-level="13.5" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#shared-ancestry-and-continuous-character-evolution"><i class="fa fa-check"></i><b>13.5</b> Shared ancestry and continuous character evolution</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#always-visualize-your-data"><i class="fa fa-check"></i><b>13.5.1</b> Always visualize your data</a></li>
<li class="chapter" data-level="13.5.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#fitting-bm-on-flower-length"><i class="fa fa-check"></i><b>13.5.2</b> Fitting BM on flower length</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html"><i class="fa fa-check"></i><b>14</b> Phylogenetic Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="14.1" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#phylogenetic-linear-models"><i class="fa fa-check"></i><b>14.1</b> Phylogenetic Linear Models</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#testing-the-hypothesis"><i class="fa fa-check"></i><b>14.1.1</b> Testing the hypothesis</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#what-happens-if-the-model-of-evolution-is-not-brownian-motion-but-something-else"><i class="fa fa-check"></i><b>14.2</b> What happens if the model of evolution is not Brownian motion but something else?</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html"><i class="fa fa-check"></i><b>15</b> Discrete variable modeling</a>
<ul>
<li class="chapter" data-level="15.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#plotting-discrete-data-on-a-phylogeny"><i class="fa fa-check"></i><b>15.1</b> Plotting discrete data on a phylogeny</a></li>
<li class="chapter" data-level="15.2" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#discrete-character-evolution"><i class="fa fa-check"></i><b>15.2</b> Discrete character evolution</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#what-is-a-q-matrix"><i class="fa fa-check"></i><b>15.2.1</b> What is a Q-matrix?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#modeling-using-ctmcs"><i class="fa fa-check"></i><b>15.3</b> Modeling using CTMCs</a>
<ul>
<li class="chapter" data-level="15.3.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#modeling-the-transition-rates-between-two-states"><i class="fa fa-check"></i><b>15.3.1</b> Modeling the transition rates between two states</a></li>
<li class="chapter" data-level="15.3.2" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#model-comparison"><i class="fa fa-check"></i><b>15.3.2</b> Model comparison</a></li>
<li class="chapter" data-level="15.3.3" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#stochastic-mapping"><i class="fa fa-check"></i><b>15.3.3</b> Stochastic mapping</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Andes Phylogenomics Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="discrete-variable-modeling" class="section level1 hasAnchor" number="15">
<h1><span class="header-section-number">Chapter 15</span> Discrete variable modeling<a href="discrete-variable-modeling.html#discrete-variable-modeling" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this section, we will work on modeling a continuous trait and understanding what Brownian motion is.</p>
<ul>
<li>In addition to the libraries loaded in the previous chapters, load <code>expm</code> and <code>corHMM</code>.</li>
<li>Ensure that the tree and dataset from the prior chapters are loaded.</li>
</ul>
<div id="plotting-discrete-data-on-a-phylogeny" class="section level2 hasAnchor" number="15.1">
<h2><span class="header-section-number">15.1</span> Plotting discrete data on a phylogeny<a href="discrete-variable-modeling.html#plotting-discrete-data-on-a-phylogeny" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There are multiple ways to plot discrete data on a phylogeny.
We will use <code>phytools</code>, but there are also tools in <code>ape</code> and <code>ggtree</code>.</p>
<p>Using <code>phytools</code> we will plot a tree with tip data as colors based on hummingbird pollination.</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="discrete-variable-modeling.html#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTree</span>(pole_tree,<span class="at">ftype=</span><span class="st">&quot;i&quot;</span>,<span class="at">offset=</span><span class="fl">0.6</span>,<span class="at">fsize=</span><span class="fl">0.3</span>)</span></code></pre></div>
<p>If you look at <code>pole_dataset</code> you see that <code>P1</code> (the category of whether butterflies visit or not)
is either 1 or 0. To the researcher this is like true/false, but the computer considers this vector to be numeric. As we did in a prior chapter (where we converted character data to numeric), we want to let the computer know this is categorical (factor) data.</p>
<pre><code>aux &lt;- as.factor(pole_dataset$P1)
aux &lt;- setNames(aux, pole_tree$tip.label)</code></pre>
<p>Now we need to create a matrix for the pie chart.</p>
<pre><code>discrete.matrix &lt;- to.matrix(aux,levels(aux))</code></pre>
<p>And finally we add these pies to the tips of the tree.
You may need to zoom in to see the labels on the tips of the tree.</p>
<pre><code>tiplabels(pie=discrete.matrix,piecol=c(&quot;blue&quot;,&quot;red&quot;),cex=0.1)</code></pre>
<p>What conclusions do you draw about the evolution of polination based on this tree?</p>
<p>For clarity add a legend to your figure.</p>
<pre><code>legend(&quot;topleft&quot;,levels(aux),pch=21,pt.bg=c(&quot;blue&quot;,&quot;red&quot;),
    pt.cex=2.2)</code></pre>
<p>Notice that the function <code>tiplabels()</code> can draw pies, meaning it can draw more than one state for a tip.</p>
</div>
<div id="discrete-character-evolution" class="section level2 hasAnchor" number="15.2">
<h2><span class="header-section-number">15.2</span> Discrete character evolution<a href="discrete-variable-modeling.html#discrete-character-evolution" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For models of discrete character evolution we will use a mathematical model called <em>Continuous-time Markov chains (CTMC)</em>. CTMCs are stochastic (random) processes that allow us to follow the evolution of a trait(s) of interest over the branches of the tree. In this example we will focus on the evolution of hummingbird pollination.</p>
<p>CTMCs are usually denoted in mathematical notation as <span class="math display">\[\{X(t), t\geq 0 \}\]</span>; the stochastic process <span class="math display">\[X(t)\]</span> follows the pollination throughout time <span class="math display">\[t\]</span>. Time is continuous and it is measured using branch length of our phylogeny. For discrete traits the process <span class="math display">\[X(t)\]</span> takes values in the natural numbers <span class="math display">\[0,1,2,...\]</span>. CTMCs have many (cool) mathematical properties, but one of the most important property is the <em>Markovian property</em> that states that the probability distribution of future states of the process conditional on both past and present states depends only upon the present state.</p>
<p>Mathematically, CTMC can is defined through a circle and arrow diagram and that gets converted into a matrix that we called the Q-matrix.</p>
<div id="what-is-a-q-matrix" class="section level3 hasAnchor" number="15.2.1">
<h3><span class="header-section-number">15.2.1</span> What is a Q-matrix?<a href="discrete-variable-modeling.html#what-is-a-q-matrix" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In mathematics, the Q-matrix is the derivative of the probability matrix (<span class="math display">\[P&#39;(t)=P(t)Q\]</span>). The elements outside of the diagonal of the <span class="math display">\[Q\]</span> matrix are transition rates and in the diagonal we have the negative sum of all the elements of the row. The row adds to zero this is because probability matrices rows add to 1 so the derivative of the constant 1 is zero.</p>
<p>Let’s stop here to think through this problem by drawing it.</p>
</div>
</div>
<div id="modeling-using-ctmcs" class="section level2 hasAnchor" number="15.3">
<h2><span class="header-section-number">15.3</span> Modeling using CTMCs<a href="discrete-variable-modeling.html#modeling-using-ctmcs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In phylogenetic comparative methods we have different types of CTMCs. Examples of these models are:</p>
<ol style="list-style-type: decimal">
<li>Evolution of tempo of a continuous trait (past lesson) -&gt; BM, OU, EB, Levy, implemented in <code>geiger</code>, <code>phylolm</code>, <code>bayou</code> R packages.</li>
<li>Evolution of the tempo between states-&gt; Markov models (Mkn) implemented in <code>phytools</code>, <code>corHMM</code></li>
<li>Evolution of the tempo between states considering other potential sources promoting evolution-&gt; Hidden state Markov models (HMM) implemented in <code>corHMM</code></li>
<li>Phylogeographic models that include dispersal or extirpation and cladogenesis -&gt; DEC, DECj, DiVA implemented in <code>Biogeobears</code></li>
<li>Diversification (speciation and extinction) that changes over time -&gt; Birth and death models implemented in <code>castor</code></li>
<li>Diversification (speciation and extinction) that changes based on clades or density -&gt; Birth and death models implemented in <code>BAMM</code>, <code>CLADS</code></li>
<li>Diversification (speciation and extinction) that changes based on state values-&gt; State-dependent diversification models for example BiSSE, HiSSE implemented in <code>hisse</code></li>
<li>Diversification (speciation and extinction) that changes based on biogeography or cladogenesis -&gt; State-dependent diversification models for example GeoSSE implemented in <code>hisse</code></li>
</ol>
<p>Learning about CTMCs is then critical for us to develop hypotheses about trait evolution, diversification.</p>
<div id="modeling-the-transition-rates-between-two-states" class="section level3 hasAnchor" number="15.3.1">
<h3><span class="header-section-number">15.3.1</span> Modeling the transition rates between two states<a href="discrete-variable-modeling.html#modeling-the-transition-rates-between-two-states" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We are going to use a CTMC model called Mk2 (Markov model with two states). Let’s take a minute to draw define the model. We are interested in the rate of evolution of hummingbird pollination in Phlox species.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="discrete-variable-modeling.html#cb109-1" aria-hidden="true" tabindex="-1"></a>data.discrete <span class="ot">&lt;-</span> <span class="fu">cbind</span>(pole_tree<span class="sc">$</span>tip.label, pole_dataset<span class="sc">$</span>P1)</span>
<span id="cb109-2"><a href="discrete-variable-modeling.html#cb109-2" aria-hidden="true" tabindex="-1"></a>mk2 <span class="ot">&lt;-</span> <span class="fu">corHMM</span>(<span class="at">phy=</span>pole_tree, <span class="at">data=</span>data.discrete, <span class="at">rate.cat=</span><span class="dv">1</span>)</span>
<span id="cb109-3"><a href="discrete-variable-modeling.html#cb109-3" aria-hidden="true" tabindex="-1"></a>mk2 <span class="co">#Let&#39;s read and interpret the results carefully</span></span>
<span id="cb109-4"><a href="discrete-variable-modeling.html#cb109-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMKmodel</span>(mk2)</span></code></pre></div>
<p>Assumption: Equal rates</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="discrete-variable-modeling.html#cb110-1" aria-hidden="true" tabindex="-1"></a>LegendAndRateMat <span class="ot">&lt;-</span> <span class="fu">getStateMat4Dat</span>(data.discrete)</span>
<span id="cb110-2"><a href="discrete-variable-modeling.html#cb110-2" aria-hidden="true" tabindex="-1"></a>RateMat <span class="ot">&lt;-</span> LegendAndRateMat<span class="sc">$</span>rate.mat</span>
<span id="cb110-3"><a href="discrete-variable-modeling.html#cb110-3" aria-hidden="true" tabindex="-1"></a>pars2equal <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb110-4"><a href="discrete-variable-modeling.html#cb110-4" aria-hidden="true" tabindex="-1"></a>constrained.mat<span class="ot">&lt;-</span><span class="fu">equateStateMatPars</span>(RateMat,pars2equal)</span>
<span id="cb110-5"><a href="discrete-variable-modeling.html#cb110-5" aria-hidden="true" tabindex="-1"></a>constrained.mat</span>
<span id="cb110-6"><a href="discrete-variable-modeling.html#cb110-6" aria-hidden="true" tabindex="-1"></a>mk2_equalrates <span class="ot">&lt;-</span> <span class="fu">corHMM</span>(<span class="at">phy=</span>pole_tree, <span class="at">data=</span>data.discrete, <span class="at">rate.cat=</span><span class="dv">1</span>, <span class="at">rate.mat=</span>constrained.mat)</span>
<span id="cb110-7"><a href="discrete-variable-modeling.html#cb110-7" aria-hidden="true" tabindex="-1"></a>mk2_equalrates</span>
<span id="cb110-8"><a href="discrete-variable-modeling.html#cb110-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMKmodel</span>(mk2_equalrates)</span></code></pre></div>
</div>
<div id="model-comparison" class="section level3 hasAnchor" number="15.3.2">
<h3><span class="header-section-number">15.3.2</span> Model comparison<a href="discrete-variable-modeling.html#model-comparison" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is important to use model comparison in this context to investigate whether the rates of evolution between the two states are equal or different. In likelihood context we use the Akaike Information Criterion known as AIC that is defined as
<span class="math display">\[AIC= -2 \log(\textrm{Likelihood evaluated in the MLE)}+ 2\times \textrm{number of parameters}\]</span>
AIC allows us to compare between models with different number of parameters because it corrects for it. We want to choose for the model with the smallest AIC. A difference of 2 or larger between two AIC values is considered as evidence in favor of a given model.</p>
<p>So knowing this, which model of evolution would you choose for this example?</p>
</div>
<div id="stochastic-mapping" class="section level3 hasAnchor" number="15.3.3">
<h3><span class="header-section-number">15.3.3</span> Stochastic mapping<a href="discrete-variable-modeling.html#stochastic-mapping" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Visualizing the history of evolution of a trait using stochastic maps.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="discrete-variable-modeling.html#cb111-1" aria-hidden="true" tabindex="-1"></a>phy <span class="ot">&lt;-</span> mk2_equalrates<span class="sc">$</span>phy</span>
<span id="cb111-2"><a href="discrete-variable-modeling.html#cb111-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span>mk2_equalrates<span class="sc">$</span>data</span>
<span id="cb111-3"><a href="discrete-variable-modeling.html#cb111-3" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> mk2_equalrates<span class="sc">$</span>solution</span>
<span id="cb111-4"><a href="discrete-variable-modeling.html#cb111-4" aria-hidden="true" tabindex="-1"></a>model[<span class="fu">is.na</span>(model)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb111-5"><a href="discrete-variable-modeling.html#cb111-5" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(model) <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fu">rowSums</span>(model)</span>
<span id="cb111-6"><a href="discrete-variable-modeling.html#cb111-6" aria-hidden="true" tabindex="-1"></a><span class="co"># run get simmap (can be plotted using phytools)</span></span>
<span id="cb111-7"><a href="discrete-variable-modeling.html#cb111-7" aria-hidden="true" tabindex="-1"></a>simmap <span class="ot">&lt;-</span> <span class="fu">makeSimmap</span>(<span class="at">tree=</span>phy, <span class="at">data=</span>data, <span class="at">model=</span>model, <span class="at">rate.cat=</span><span class="dv">1</span>, <span class="at">nSim=</span><span class="dv">10</span>, <span class="at">nCores=</span><span class="dv">1</span>)</span>
<span id="cb111-8"><a href="discrete-variable-modeling.html#cb111-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-9"><a href="discrete-variable-modeling.html#cb111-9" aria-hidden="true" tabindex="-1"></a><span class="co"># we import phytools plotSimmap for plotting</span></span>
<span id="cb111-10"><a href="discrete-variable-modeling.html#cb111-10" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb111-11"><a href="discrete-variable-modeling.html#cb111-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plotSimmap</span>(simmap[[<span class="dv">1</span>]], <span class="at">fsize=</span>.<span class="dv">4</span>)</span>
<span id="cb111-12"><a href="discrete-variable-modeling.html#cb111-12" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">&quot;black&quot;</span>,<span class="st">&quot;red&quot;</span>),<span class="dv">0</span><span class="sc">:</span><span class="dv">1</span>)</span>
<span id="cb111-13"><a href="discrete-variable-modeling.html#cb111-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb111-14"><a href="discrete-variable-modeling.html#cb111-14" aria-hidden="true" tabindex="-1"></a><span class="fu">densityTree</span>(simmap,<span class="at">method=</span><span class="st">&quot;plotSimmap&quot;</span>,<span class="at">lwd=</span><span class="dv">8</span>,<span class="at">nodes=</span><span class="st">&quot;intermediate&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">170</span>),<span class="at">colors=</span>colors, <span class="at">compute.consensus=</span><span class="cn">FALSE</span>, <span class="at">fsize=</span><span class="fl">0.2</span>, <span class="at">show.axis=</span><span class="cn">FALSE</span>, <span class="at">fix.depth=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<p>,
### Modeling hidden states</p>
<p>Often, the transitions between to states are associated with other changes in the history of the trait itself. Some of those changes we can measure i.e., polyploidy vs. breeding systems, but some others are hard to quantify or hard to measure. Because many things could have happen that are linked to transitions between states we need to model using what we call hidden states.</p>
<p>The hidden states model is a mathematical tool that allow us to parse the main effect of the trait of interest, by removing other potential sources of evolution in the transitons.</p>
<p>We will stop here and draw what these hidden states are!</p>
<p>How do we implement them in R?
Two different assumptions</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="discrete-variable-modeling.html#cb112-1" aria-hidden="true" tabindex="-1"></a>hmm2_2_equal <span class="ot">&lt;-</span> <span class="fu">corHMM</span>(<span class="at">phy =</span> pole_tree, <span class="at">data =</span> data.discrete, <span class="at">rate.cat =</span> <span class="dv">2</span>, <span class="at">model =</span> <span class="st">&quot;SYM&quot;</span>, <span class="at">get.tip.states =</span> <span class="cn">TRUE</span>)</span>
<span id="cb112-2"><a href="discrete-variable-modeling.html#cb112-2" aria-hidden="true" tabindex="-1"></a>hmm2_2_equal <span class="co">#Let&#39;s read and interpret the results carefully</span></span>
<span id="cb112-3"><a href="discrete-variable-modeling.html#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMKmodel</span>(hmm2_2_equal)</span>
<span id="cb112-4"><a href="discrete-variable-modeling.html#cb112-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-5"><a href="discrete-variable-modeling.html#cb112-5" aria-hidden="true" tabindex="-1"></a>hmm2_2_different<span class="ot">&lt;-</span> <span class="fu">corHMM</span>(<span class="at">phy =</span> pole_tree, <span class="at">data =</span> data.discrete, <span class="at">rate.cat =</span> <span class="dv">2</span>, <span class="at">model =</span> <span class="st">&quot;ARD&quot;</span>, <span class="at">get.tip.states =</span> <span class="cn">TRUE</span>)</span>
<span id="cb112-6"><a href="discrete-variable-modeling.html#cb112-6" aria-hidden="true" tabindex="-1"></a>hmm2_2_different <span class="co">#Let&#39;s read and interpret the results carefully</span></span>
<span id="cb112-7"><a href="discrete-variable-modeling.html#cb112-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMKmodel</span>(hmm2_2_different)</span></code></pre></div>
<p>What is a better? What is the overall best model??</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="phylogenetic-generalized-linear-models.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/15-macrodiscrete.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
