<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Introduction to macroevolution modeling | Andes Phylogenomics Workshop</title>
  <meta name="description" content="Step-by-step guide to a project in phylogenomics." />
  <meta name="generator" content="bookdown 0.34 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Introduction to macroevolution modeling | Andes Phylogenomics Workshop" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="Step-by-step guide to a project in phylogenomics." />
  <meta name="github-repo" content="rachelss/Andes_workshop_codebook" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Introduction to macroevolution modeling | Andes Phylogenomics Workshop" />
  
  <meta name="twitter:description" content="Step-by-step guide to a project in phylogenomics." />
  

<meta name="author" content="Rachel Schwartz and Rosana Zenil-Ferguson" />


<meta name="date" content="2023-06-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="what-is-bayesian-statistics.html"/>
<link rel="next" href="phylogenetic-generalized-linear-models.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Andes Phylogenomics Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About this workshop</a></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a>
<ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#case-studies"><i class="fa fa-check"></i><b>2.2</b> Case Studies</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#course-objectives"><i class="fa fa-check"></i><b>2.3</b> Course Objectives</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="data-and-analysis.html"><a href="data-and-analysis.html"><i class="fa fa-check"></i><b>3</b> Data and analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="data-and-analysis.html"><a href="data-and-analysis.html#genomic-data"><i class="fa fa-check"></i><b>3.1</b> Genomic data</a></li>
<li class="chapter" data-level="3.2" data-path="data-and-analysis.html"><a href="data-and-analysis.html#phylogenies"><i class="fa fa-check"></i><b>3.2</b> Phylogenies</a></li>
<li class="chapter" data-level="3.3" data-path="data-and-analysis.html"><a href="data-and-analysis.html#comparative-analyses"><i class="fa fa-check"></i><b>3.3</b> Comparative analyses</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="trait-data.html"><a href="trait-data.html"><i class="fa fa-check"></i><b>4</b> Trait Data</a>
<ul>
<li class="chapter" data-level="4.1" data-path="trait-data.html"><a href="trait-data.html#getting-started-with-r"><i class="fa fa-check"></i><b>4.1</b> Getting started with R</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="trait-data.html"><a href="trait-data.html#option-1-log-in-to-the-server"><i class="fa fa-check"></i><b>4.1.1</b> Option 1: Log in to the server</a></li>
<li class="chapter" data-level="4.1.2" data-path="trait-data.html"><a href="trait-data.html#option-2-work-on-your-personal-computer"><i class="fa fa-check"></i><b>4.1.2</b> Option 2: Work on your personal computer</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="trait-data.html"><a href="trait-data.html#load-and-set-up-data"><i class="fa fa-check"></i><b>4.2</b> Load and set up data</a></li>
<li class="chapter" data-level="4.3" data-path="trait-data.html"><a href="trait-data.html#beginning-an-analysis"><i class="fa fa-check"></i><b>4.3</b> Beginning an analysis</a></li>
<li class="chapter" data-level="4.4" data-path="trait-data.html"><a href="trait-data.html#data-visualization"><i class="fa fa-check"></i><b>4.4</b> Data Visualization</a></li>
<li class="chapter" data-level="4.5" data-path="trait-data.html"><a href="trait-data.html#statistical-tests-of-relationships"><i class="fa fa-check"></i><b>4.5</b> Statistical tests of relationships</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="computational-skills.html"><a href="computational-skills.html"><i class="fa fa-check"></i><b>5</b> Computational skills</a>
<ul>
<li class="chapter" data-level="5.1" data-path="computational-skills.html"><a href="computational-skills.html#basic-navigation"><i class="fa fa-check"></i><b>5.1</b> Basic navigation</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="getting-your-data.html"><a href="getting-your-data.html"><i class="fa fa-check"></i><b>6</b> Getting your data</a>
<ul>
<li class="chapter" data-level="6.1" data-path="getting-your-data.html"><a href="getting-your-data.html#repeating-analyses"><i class="fa fa-check"></i><b>6.1</b> Repeating analyses</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="tree-thinking-quiz.html"><a href="tree-thinking-quiz.html"><i class="fa fa-check"></i><b>7</b> Tree thinking quiz</a></li>
<li class="chapter" data-level="8" data-path="target-capture-data.html"><a href="target-capture-data.html"><i class="fa fa-check"></i><b>8</b> Target capture data</a>
<ul>
<li class="chapter" data-level="8.1" data-path="target-capture-data.html"><a href="target-capture-data.html#background"><i class="fa fa-check"></i><b>8.1</b> Background</a></li>
<li class="chapter" data-level="8.2" data-path="target-capture-data.html"><a href="target-capture-data.html#running-hybpiper-to-assemble-genes"><i class="fa fa-check"></i><b>8.2</b> Running HybPiper to assemble genes</a></li>
<li class="chapter" data-level="8.3" data-path="target-capture-data.html"><a href="target-capture-data.html#process-outputs"><i class="fa fa-check"></i><b>8.3</b> Process outputs</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="target-capture-data.html"><a href="target-capture-data.html#viewing-information"><i class="fa fa-check"></i><b>8.3.1</b> Viewing information</a></li>
<li class="chapter" data-level="8.3.2" data-path="target-capture-data.html"><a href="target-capture-data.html#obtain-gene-data"><i class="fa fa-check"></i><b>8.3.2</b> Obtain gene data</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="target-capture-data.html"><a href="target-capture-data.html#visual-inspection-of-data"><i class="fa fa-check"></i><b>8.4</b> Visual inspection of data</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="results.html"><a href="results.html"><i class="fa fa-check"></i><b>9</b> Results</a>
<ul>
<li class="chapter" data-level="9.1" data-path="results.html"><a href="results.html#alignment"><i class="fa fa-check"></i><b>9.1</b> Alignment</a></li>
<li class="chapter" data-level="9.2" data-path="results.html"><a href="results.html#view-in-r"><i class="fa fa-check"></i><b>9.2</b> View in R</a></li>
<li class="chapter" data-level="9.3" data-path="results.html"><a href="results.html#concatenated-data-analysis"><i class="fa fa-check"></i><b>9.3</b> Concatenated data analysis</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="results.html"><a href="results.html#building-trees-with-iqtree"><i class="fa fa-check"></i><b>9.3.1</b> Building trees with IQTree</a></li>
<li class="chapter" data-level="9.3.2" data-path="results.html"><a href="results.html#viewing-trees"><i class="fa fa-check"></i><b>9.3.2</b> Viewing trees</a></li>
<li class="chapter" data-level="9.3.3" data-path="results.html"><a href="results.html#tree-for-all-data-using-output-of-individual-hybpiper-runs"><i class="fa fa-check"></i><b>9.3.3</b> Tree for all data using output of individual hybpiper runs</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="results.html"><a href="results.html#species-tree-analyses"><i class="fa fa-check"></i><b>9.4</b> Species tree analyses</a></li>
<li class="chapter" data-level="9.5" data-path="results.html"><a href="results.html#support-for-relationships"><i class="fa fa-check"></i><b>9.5</b> Support for relationships</a></li>
<li class="chapter" data-level="9.6" data-path="results.html"><a href="results.html#mrbayes"><i class="fa fa-check"></i><b>9.6</b> MrBayes</a></li>
<li class="chapter" data-level="9.7" data-path="results.html"><a href="results.html#what-do-your-trees-tell-you"><i class="fa fa-check"></i><b>9.7</b> What do your trees tell you?</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="dates-and-evolution.html"><a href="dates-and-evolution.html"><i class="fa fa-check"></i><b>10</b> Dates and Evolution</a>
<ul>
<li class="chapter" data-level="10.1" data-path="dates-and-evolution.html"><a href="dates-and-evolution.html#dates-times-on-trees"><i class="fa fa-check"></i><b>10.1</b> Dates / Times on Trees</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html"><i class="fa fa-check"></i><b>11</b> Understanding the likelihood function</a>
<ul>
<li class="chapter" data-level="11.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#background-1"><i class="fa fa-check"></i><b>11.1</b> Background</a></li>
<li class="chapter" data-level="11.2" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-model"><i class="fa fa-check"></i><b>11.2</b> The model</a></li>
<li class="chapter" data-level="11.3" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-likelihood-function"><i class="fa fa-check"></i><b>11.3</b> The likelihood function</a></li>
<li class="chapter" data-level="11.4" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#a-mini-example-with-a-sample-size-of-two"><i class="fa fa-check"></i><b>11.4</b> A mini example with a sample size of two</a></li>
<li class="chapter" data-level="11.5" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#the-bird-alarm-example"><i class="fa fa-check"></i><b>11.5</b> The bird alarm example</a>
<ul>
<li class="chapter" data-level="11.5.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#calculate-the-mle-for-dl-and-f-and-the-likelihood-value-evaluated-at-the-mle"><i class="fa fa-check"></i><b>11.5.1</b> Calculate the MLE for DL and F and the likelihood value evaluated at the MLE</a></li>
</ul></li>
<li class="chapter" data-level="11.6" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#where-is-the-evidence-for-different-behavior"><i class="fa fa-check"></i><b>11.6</b> Where is the evidence for different behavior?</a>
<ul>
<li class="chapter" data-level="11.6.1" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#plot-the-likelihood-for-dead-leaf-gleaners"><i class="fa fa-check"></i><b>11.6.1</b> Plot the likelihood for dead-leaf gleaners</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="understanding-the-likelihood-function.html"><a href="understanding-the-likelihood-function.html#are-the-freezing-times-different-for-the-two-groups"><i class="fa fa-check"></i><b>11.7</b> Are the freezing times different for the two groups?</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html"><i class="fa fa-check"></i><b>12</b> What is Bayesian statistics?</a>
<ul>
<li class="chapter" data-level="12.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#what-about-all-the-fighting-of-frequentists-vs.-bayesians"><i class="fa fa-check"></i><b>12.1</b> What about all the fighting of “frequentists” vs. “Bayesians”</a></li>
<li class="chapter" data-level="12.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#when-do-we-choose-bayesian-statistics"><i class="fa fa-check"></i><b>12.2</b> When do we choose Bayesian statistics?</a></li>
<li class="chapter" data-level="12.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#basic-concepts-of-bayesian-statistics"><i class="fa fa-check"></i><b>12.3</b> Basic concepts of Bayesian Statistics</a></li>
<li class="chapter" data-level="12.4" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#example-online-dating"><i class="fa fa-check"></i><b>12.4</b> Example: Online dating</a>
<ul>
<li class="chapter" data-level="12.4.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#pab-is-difficult-to-estimate"><i class="fa fa-check"></i><b>12.4.1</b> <span class="math inline">\(P(A|B)\)</span> is difficult to estimate</a></li>
<li class="chapter" data-level="12.4.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#coming-back-to-the-initial-idea-of-how-we-define-bayesian-statistics"><i class="fa fa-check"></i><b>12.4.2</b> Coming back to the initial idea of how we define Bayesian statistics</a></li>
<li class="chapter" data-level="12.4.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-random-variable"><i class="fa fa-check"></i><b>12.4.3</b> The random variable</a></li>
<li class="chapter" data-level="12.4.4" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-prior-distribution"><i class="fa fa-check"></i><b>12.4.4</b> The prior distribution</a></li>
<li class="chapter" data-level="12.4.5" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-data-and-their-distribution"><i class="fa fa-check"></i><b>12.4.5</b> The Data and their distribution</a></li>
</ul></li>
<li class="chapter" data-level="12.5" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-likelihood-function-1"><i class="fa fa-check"></i><b>12.5</b> The likelihood function</a></li>
<li class="chapter" data-level="12.6" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-update-the-posterior-distribution"><i class="fa fa-check"></i><b>12.6</b> The update! The posterior distribution</a></li>
<li class="chapter" data-level="12.7" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#optional-section-doing-your-first-mcmc"><i class="fa fa-check"></i><b>12.7</b> Optional section: Doing your first MCMC</a>
<ul>
<li class="chapter" data-level="12.7.1" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-proposal-function"><i class="fa fa-check"></i><b>12.7.1</b> The proposal function</a></li>
<li class="chapter" data-level="12.7.2" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#the-metropolis-hastings"><i class="fa fa-check"></i><b>12.7.2</b> The Metropolis-Hastings</a></li>
<li class="chapter" data-level="12.7.3" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#visualizing-the-mcmc-as-a-way-to-find-out-convergence"><i class="fa fa-check"></i><b>12.7.3</b> Visualizing the MCMC as a way to find out convergence</a></li>
</ul></li>
<li class="chapter" data-level="12.8" data-path="what-is-bayesian-statistics.html"><a href="what-is-bayesian-statistics.html#changing-the-proposal"><i class="fa fa-check"></i><b>12.8</b> Changing the proposal</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html"><i class="fa fa-check"></i><b>13</b> Introduction to macroevolution modeling</a>
<ul>
<li class="chapter" data-level="13.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#reading-dataset-and-phylogenetic-tree"><i class="fa fa-check"></i><b>13.1</b> Reading dataset and phylogenetic tree</a></li>
<li class="chapter" data-level="13.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-a-phylogenetic-tree-with-data"><i class="fa fa-check"></i><b>13.2</b> Plotting a phylogenetic tree with data</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-continuous-data"><i class="fa fa-check"></i><b>13.2.1</b> Plotting continuous data</a></li>
<li class="chapter" data-level="13.2.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#plotting-discrete-and-continuous-data"><i class="fa fa-check"></i><b>13.2.2</b> Plotting discrete and continuous data</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#introduction-to-brownian-motion"><i class="fa fa-check"></i><b>13.3</b> Introduction to Brownian Motion</a></li>
<li class="chapter" data-level="13.4" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#brownian-motion-in-phylogenetic-comparative-methods"><i class="fa fa-check"></i><b>13.4</b> Brownian motion in phylogenetic comparative methods</a></li>
<li class="chapter" data-level="13.5" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#shared-ancestry-and-continuous-character-evolution"><i class="fa fa-check"></i><b>13.5</b> Shared ancestry and continuous character evolution</a>
<ul>
<li class="chapter" data-level="13.5.1" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#always-visualize-your-data"><i class="fa fa-check"></i><b>13.5.1</b> Always visualize your data</a></li>
<li class="chapter" data-level="13.5.2" data-path="introduction-to-macroevolution-modeling.html"><a href="introduction-to-macroevolution-modeling.html#fitting-bm-on-flower-length"><i class="fa fa-check"></i><b>13.5.2</b> Fitting BM on flower length</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="14" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html"><i class="fa fa-check"></i><b>14</b> Phylogenetic Generalized Linear Models</a>
<ul>
<li class="chapter" data-level="14.1" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#phylogenetic-linear-models"><i class="fa fa-check"></i><b>14.1</b> Phylogenetic Linear Models</a>
<ul>
<li class="chapter" data-level="14.1.1" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#testing-the-hypothesis"><i class="fa fa-check"></i><b>14.1.1</b> Testing the hypothesis</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="phylogenetic-generalized-linear-models.html"><a href="phylogenetic-generalized-linear-models.html#what-happens-if-the-model-of-evolution-is-not-brownian-motion-but-something-else"><i class="fa fa-check"></i><b>14.2</b> What happens if the model of evolution is not Brownian motion but something else?</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html"><i class="fa fa-check"></i><b>15</b> Discrete variable modeling</a>
<ul>
<li class="chapter" data-level="15.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#plotting-discrete-data-on-a-phylogeny"><i class="fa fa-check"></i><b>15.1</b> Plotting discrete data on a phylogeny</a></li>
<li class="chapter" data-level="15.2" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#discrete-character-evolution"><i class="fa fa-check"></i><b>15.2</b> Discrete character evolution</a>
<ul>
<li class="chapter" data-level="15.2.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#what-is-a-q-matrix"><i class="fa fa-check"></i><b>15.2.1</b> What is a Q-matrix?</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#modeling-using-ctmcs"><i class="fa fa-check"></i><b>15.3</b> Modeling using CTMCs</a>
<ul>
<li class="chapter" data-level="15.3.1" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#modeling-the-transition-rates-between-two-states"><i class="fa fa-check"></i><b>15.3.1</b> Modeling the transition rates between two states</a></li>
<li class="chapter" data-level="15.3.2" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#model-comparison"><i class="fa fa-check"></i><b>15.3.2</b> Model comparison</a></li>
<li class="chapter" data-level="15.3.3" data-path="discrete-variable-modeling.html"><a href="discrete-variable-modeling.html#stochastic-mapping"><i class="fa fa-check"></i><b>15.3.3</b> Stochastic mapping</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Andes Phylogenomics Workshop</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-to-macroevolution-modeling" class="section level1 hasAnchor" number="13">
<h1><span class="header-section-number">Chapter 13</span> Introduction to macroevolution modeling<a href="introduction-to-macroevolution-modeling.html#introduction-to-macroevolution-modeling" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this section, we will work on modeling a continuous trait and understanding what Brownian motion is.</p>
<p>Start by loading the libraries <code>ape</code>, <code>geiger</code>, <code>phytools</code>, <code>nlme</code>, and <code>phylolm</code>.</p>
<div id="reading-dataset-and-phylogenetic-tree" class="section level2 hasAnchor" number="13.1">
<h2><span class="header-section-number">13.1</span> Reading dataset and phylogenetic tree<a href="introduction-to-macroevolution-modeling.html#reading-dataset-and-phylogenetic-tree" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Data is a simplify version of the data and models we published in
Landis, J.B., Bell, C.D., Hernandez, M., Zenil-Ferguson, R., McCarthy, E.W., Soltis, D.E. and Soltis, P.S., 2018. Evolution of floral traits and impact of reproductive mode on diversification in the phlox family (Polemoniaceae).Molecular phylogenetics and evolution, 127, pp.878-890.
We have a tree with 165 tips from the Polimoniaceae family (Phlox). They grow in California. We are interested in the rate of evolution of flower length.</p>
<p>Read in the <code>pole_tree.nex</code> file from the shared folder using the <code>read.nexus</code> function.</p>
<ul>
<li>Read the dataset of traits and pollinators visiting the flowers on the tips of the tree. This file is <code>pole_dataset.csv</code> in the shared folder.</li>
<li>For many phylogenetic models we need to make sure that the row names match the tip labels on the tree. Assign the tip labels of the <code>pole_tree</code> dataset to <code>rownames(pole_dataset)</code>.</li>
<li>View the dataset.</li>
</ul>
<p>This assumes that the dataset is already in the same order as the tree file.</p>
<p>You should see columns for</p>
<ul>
<li>Length- flower length (Response)</li>
<li>P1- butterflies visit or not (Predictor)</li>
<li>P2- hawkmoths visit or not (Predictor)</li>
<li>P3- bee flies visit or not (Predictor)</li>
<li>P4- bees visit or not (Predictor)</li>
</ul>
</div>
<div id="plotting-a-phylogenetic-tree-with-data" class="section level2 hasAnchor" number="13.2">
<h2><span class="header-section-number">13.2</span> Plotting a phylogenetic tree with data<a href="introduction-to-macroevolution-modeling.html#plotting-a-phylogenetic-tree-with-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li>Plot this tree</li>
</ul>
<p>An easy way to visualize phylogenies that are small is by plotting them in a rectangular fashion
as we have done, but when you have a lot of tips, circular phylogenies can be more useful.
You can use the <code>type</code> argument to plot to change the plot style.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="introduction-to-macroevolution-modeling.html#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pole_tree,<span class="at">cex=</span><span class="fl">0.2</span>,<span class="at">type=</span><span class="st">&quot;fan&quot;</span>)</span></code></pre></div>
<div id="plotting-continuous-data" class="section level3 hasAnchor" number="13.2.1">
<h3><span class="header-section-number">13.2.1</span> Plotting continuous data<a href="introduction-to-macroevolution-modeling.html#plotting-continuous-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Jacob Landis measured the flower width and length and observed which pollinators come visit the flowers. Jacob was interested in understanding if we can predict flower shape based on pollinator type. For example, flowers that hummingbirds visit might be longer and narrower than flowers that are visited by bees.</p>
<p>I wanted to create a model where the response is the flower shape and the predictor is what type of pollinator visits the flower.</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="introduction-to-macroevolution-modeling.html#cb88-1" aria-hidden="true" tabindex="-1"></a>lengthplot <span class="ot">&lt;-</span> <span class="fu">setNames</span>(pole_dataset<span class="sc">$</span>Length,<span class="fu">rownames</span>(pole_dataset))</span></code></pre></div>
<p>Stop to see what this line of code is doing.</p>
<pre><code>plotTree.barplot(tree=pole_tree,
                 x=lengthplot,
                 args.plotTree=list(fsize=0.2,ftype=&quot;i&quot;),
                 args.barplot=list(xlab=&quot;Flower length (cm)&quot;))</code></pre>
<p>Stop and see - what do you notice? Practice your tree-thinking skills.</p>
</div>
<div id="plotting-discrete-and-continuous-data" class="section level3 hasAnchor" number="13.2.2">
<h3><span class="header-section-number">13.2.2</span> Plotting discrete and continuous data<a href="introduction-to-macroevolution-modeling.html#plotting-discrete-and-continuous-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We would like to show on this plot which flowers are hummingbird pollinated.
We can do this by creating a new column in the data and we’ll put a color in it based on the information in column <code>P1</code>.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="introduction-to-macroevolution-modeling.html#cb90-1" aria-hidden="true" tabindex="-1"></a>pole_dataset<span class="sc">$</span>mycolors <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(pole_dataset<span class="sc">$</span>P1<span class="sc">==</span><span class="dv">1</span>, <span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p>Now plot the tree</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="introduction-to-macroevolution-modeling.html#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTree.barplot</span>(<span class="at">tree=</span>pole_tree,</span>
<span id="cb91-2"><a href="introduction-to-macroevolution-modeling.html#cb91-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">x=</span>lengthplot,</span>
<span id="cb91-3"><a href="introduction-to-macroevolution-modeling.html#cb91-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">args.plotTree=</span><span class="fu">list</span>(<span class="at">ftype=</span><span class="st">&quot;off&quot;</span>),</span>
<span id="cb91-4"><a href="introduction-to-macroevolution-modeling.html#cb91-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">args.barplot=</span><span class="fu">list</span>(<span class="at">col=</span>pole_dataset<span class="sc">$</span>mycolors,<span class="at">border=</span>pole_dataset<span class="sc">$</span>mycolors,<span class="at">xlab=</span><span class="st">&quot;&quot;</span>))</span>
<span id="cb91-5"><a href="introduction-to-macroevolution-modeling.html#cb91-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="at">x=</span><span class="dv">3</span>, <span class="at">y=</span><span class="dv">100</span>,<span class="at">legend=</span><span class="fu">c</span>(<span class="st">&quot;Hummingbird pollinated&quot;</span>,<span class="st">&quot;Other pollination&quot;</span>), <span class="at">pch=</span><span class="dv">22</span>, <span class="at">cex=</span><span class="fl">0.5</span>, <span class="at">pt.bg=</span><span class="fu">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>),</span>
<span id="cb91-6"><a href="introduction-to-macroevolution-modeling.html#cb91-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">box.col=</span><span class="st">&quot;transparent&quot;</span>)</span></code></pre></div>
<p>Stop and see, what kind of (macroevolution) questions can you ask after seeing this?</p>
</div>
</div>
<div id="introduction-to-brownian-motion" class="section level2 hasAnchor" number="13.3">
<h2><span class="header-section-number">13.3</span> Introduction to Brownian Motion<a href="introduction-to-macroevolution-modeling.html#introduction-to-brownian-motion" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Probability approaches have proven really useful to recreate evolutionary history across species. Although some of these models can be simplistic, they can be useful to establish some basic scenarios at which traits can evolve.</p>
<p>One of these simple, yet important models is Brownian motion (sometimes called Wiener process). The movement itself was first described by a botanist(!!!) Robert Brown (1827) who wanted to describe how pollen of Clarkia pulchella moved in water. The mathematical model was actually described by Albert Einstein in 1905.</p>
<p>Brownian motion is the result of many small forces that are added to create movement. My postdoc advisor Luke Harmon, says that it is hard for him to imagine pollen in fluid so he uses the analogy of a ball being bounced around in a large stadium as a great visual for using Brownian motion model (BM) to model the movement of the ball. It is the many small forces and directions of the people in the stadium which ultimately moves the ball all across the place.</p>
<p>Mathematically, Brownian motion is a stochastic process. This means that it is a probability model that occurs as time advances. We denote stochastic processes as <span class="math inline">\({X(t): t\geq 0}\)</span>, meaning a random variable over time, when time is greater or equal than zero. Brownian motion has three important properties:</p>
<ol style="list-style-type: decimal">
<li>The expected value of the process as time advances is the same value at which the process started. In mathematical terms <span class="math inline">\(E[X(t)]=X(0).\)</span></li>
<li>In each successive and non-overlapping time interval the process is independent. This iproperty is what we call in probability a Markovian property. In mathematical terms, we have two intervals of time <span class="math inline">\((0,t)\)</span> and <span class="math inline">\((t, t+s)\)</span>, then <span class="math inline">\([X(t)-X(0)]\)</span> is independent from <span class="math inline">\([X(t+s)-X(t))]\)</span>.</li>
<li>The process at time <span class="math inline">\(t\)</span>, that is <span class="math inline">\(X(t)\)</span> has a Normal probability distribution. This normal probability distribution has as mean value, the value at which the process started and a variance that is a function of time. In mathematical terms <span class="math inline">\(X(t)\sim N(X(0),\sigma^2t).\)</span></li>
</ol>
</div>
<div id="brownian-motion-in-phylogenetic-comparative-methods" class="section level2 hasAnchor" number="13.4">
<h2><span class="header-section-number">13.4</span> Brownian motion in phylogenetic comparative methods<a href="introduction-to-macroevolution-modeling.html#brownian-motion-in-phylogenetic-comparative-methods" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In phylogenetic comparative methods, we are assuming then that an observation of a phenotype (trait) for species A <span class="math inline">\(x_A\)</span> is then representative of the average of the population after some time. Again, this is an assumption, and as every assumption, we should question this when it comes to real data.</p>
<p>There are many traits that could be model as BM, think about body mass, brain size, foot length, all those continuous. How appropriate is to model them as BM, totally depends on the scale, and how much we believe that the trait “randomly wanders” through time. There exists other models who put a limit on how much the trait wanders, or change rapidly away from having the same expected value altogether.</p>
<p>For the purposes of this lecture, we are interested then in understanding the evolution of trait <span class="math inline">\(X(t)\)</span> (notice is a random variable over time) for two species for which we have observed values of the trait <span class="math inline">\((x_A,x_B)\)</span>. These two species are not independent, they have a shared evolutionary history dictated by a phylogenetic tree. This is a key point, in the past, all of your samples have been INDEPENDENT and identically distributed. The independent assumption is gone, so how do we go about acknowledging the shared evolutionary history?</p>
<p>Using BM model we know that
<span class="math display">\[x_A\sim N(\mu_0, \sigma^2 (t_1+t_2))\]</span>
and that
<span class="math display">\[x_B\sim N(\mu_0,\sigma^2(t_1+t_3))\]</span></p>
<p>with <span class="math inline">\(\mu_0=x(0)\)</span>. Since <span class="math inline">\((x_A, x_B)\)</span> are not independent from each other and share branch <span class="math inline">\((t_1)\)</span> in the phylogeny we can think of each value in the tips as the result of a the sum of two evolutionary changes
1. <span class="math inline">\(x_A=\Delta x_1+\Delta x_2\)</span> (the change of trait value during <span class="math inline">\(t_1\)</span> plus the change in trait value during the <span class="math inline">\(t_2\)</span> branch)
2. <span class="math inline">\(x_B=\Delta x_1+\Delta x_3\)</span> (the change of trait value during <span class="math inline">\(t_1\)</span> plus the change in trait value during the <span class="math inline">\(t_3\)</span> branch)</p>
<p>Because of the Markovian property of BM (property 2 in the introduction). We know that the change <span class="math inline">\(\delta x_2\)</span> is independent from the change in <span class="math inline">\(\delta x_1\)</span>. Also, the only shared change for the two species is whatever happened during <span class="math inline">\(t_1\)</span>. Therefore, the shared evolutionary history for the trait for species A and B is
<span class="math display">\[cov(x_A, x_B)= var(\Delta x_1)=\sigma^2 t_1\]</span>
This covariance equation is the most important and key result used in macroevolution. Extensions of these covariance represent other popular models of continuous variables (e.g. OU, EB, Lévy).</p>
</div>
<div id="shared-ancestry-and-continuous-character-evolution" class="section level2 hasAnchor" number="13.5">
<h2><span class="header-section-number">13.5</span> Shared ancestry and continuous character evolution<a href="introduction-to-macroevolution-modeling.html#shared-ancestry-and-continuous-character-evolution" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Calculate the C matrix using the vcvPhylo() function</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="introduction-to-macroevolution-modeling.html#cb92-1" aria-hidden="true" tabindex="-1"></a>C <span class="ot">&lt;-</span> <span class="fu">vcvPhylo</span>(pole_tree, <span class="at">anc.nodes=</span><span class="cn">FALSE</span>) </span></code></pre></div>
<ul>
<li>What is this the covaraince matrix of? How to interpret variance terms?</li>
<li>Check the dimensions. What is this matrix C composed of?</li>
<li>Why is it called vcvPhylo?</li>
</ul>
<div id="always-visualize-your-data" class="section level3 hasAnchor" number="13.5.1">
<h3><span class="header-section-number">13.5.1</span> Always visualize your data<a href="introduction-to-macroevolution-modeling.html#always-visualize-your-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="introduction-to-macroevolution-modeling.html#cb93-1" aria-hidden="true" tabindex="-1"></a>flower.length <span class="ot">&lt;-</span> pole_dataset<span class="sc">$</span>Length</span>
<span id="cb93-2"><a href="introduction-to-macroevolution-modeling.html#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(flower.length) <span class="ot">&lt;-</span> pole_tree<span class="sc">$</span>tip.label</span>
<span id="cb93-3"><a href="introduction-to-macroevolution-modeling.html#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(flower.length)</span></code></pre></div>
<p>What do you see here? do you think Brownian motion could be a good model for evolution for flower length?</p>
<p>Sometimes to better model BM in traits that can only be positive we often log-transform.</p>
<pre><code>log.flowerlength &lt;- log(flower.length)
hist(log.flowerlength)</code></pre>
<p>However, it is much harder to interpret log-scale.</p>
</div>
<div id="fitting-bm-on-flower-length" class="section level3 hasAnchor" number="13.5.2">
<h3><span class="header-section-number">13.5.2</span> Fitting BM on flower length<a href="introduction-to-macroevolution-modeling.html#fitting-bm-on-flower-length" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We are interested in calculating the the amount of evolution using Brownian motion for flower length. We will do this with the function <code>fitContinuous()</code> from package <code>geiger</code>.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="introduction-to-macroevolution-modeling.html#cb95-1" aria-hidden="true" tabindex="-1"></a>fitted.BM <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(pole_tree,flower.length, <span class="at">model=</span><span class="st">&quot;BM&quot;</span>)</span>
<span id="cb95-2"><a href="introduction-to-macroevolution-modeling.html#cb95-2" aria-hidden="true" tabindex="-1"></a>fitted.BM</span></code></pre></div>
<p>Stop and Interpret the output of <code>fitContinuous</code></p>
<ol style="list-style-type: decimal">
<li>Type <code>fitted.BM$opt</code> - explore a little bit what this object has</li>
<li>What is <code>fitted.BM$opt$z0</code> telling us about the Brownian motion process?</li>
<li>What is <code>fitted.BM$opt$sigsq</code> telling us about the Brownian motion process?</li>
<li>What is the log-likelihood?</li>
</ol>
<p>If you assume no phylogenetic tree and complete independence in the data, what happens with the parameters?</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="introduction-to-macroevolution-modeling.html#cb96-1" aria-hidden="true" tabindex="-1"></a>fitted.Normal <span class="ot">&lt;-</span> <span class="fu">fitContinuous</span>(pole_tree,flower.length, <span class="at">model=</span><span class="st">&quot;white&quot;</span>)</span>
<span id="cb96-2"><a href="introduction-to-macroevolution-modeling.html#cb96-2" aria-hidden="true" tabindex="-1"></a>fitted.Normal</span></code></pre></div>
<p>Which model is better? (Hint: Think about model selection, but also argue about the importance of considering the phylogeny)</p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="what-is-bayesian-statistics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="phylogenetic-generalized-linear-models.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/13-macroBM.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["bookdown-demo.pdf", "bookdown-demo.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
